From: Jinoh Kang <jinoh.kang.kr@gmail.com>
Subject: [PATCH 3/6] ntdll/tests: Add tests for NtSetInformationVirtualMemory.
Message-Id: <51b84edf-78d6-8ace-b1a5-94eb4ea0d39c@gmail.com>
Date: Thu, 20 Jan 2022 02:29:15 +0900
In-Reply-To: <d8105173-f9a2-317e-1c27-ddd290627f7a@gmail.com>
References: <d8105173-f9a2-317e-1c27-ddd290627f7a@gmail.com>

Signed-off-by: Jinoh Kang <jinoh.kang.kr@gmail.com>
---

Notes:
    v2 -> v3: Guard wow64 status code tests with todo_wine
    v4 -> v5: fix one wow64 test item
    v5 -> v6: use ULongToPtr/PtrToUlong macro; edit subject

 dlls/ntdll/tests/virtual.c | 164 +++++++++++++++++++++++++++++++++++++
 1 file changed, 164 insertions(+)

diff --git a/dlls/ntdll/tests/virtual.c b/dlls/ntdll/tests/virtual.c
index 11111111111..11111111111 100644
--- a/dlls/ntdll/tests/virtual.c
+++ b/dlls/ntdll/tests/virtual.c
@@ -41,6 +41,9 @@ static NTSTATUS (WINAPI *pNtAllocateVirtualMemoryEx)(HANDLE, PVOID *, SIZE_T *,
                                                      MEM_EXTENDED_PARAMETER *, ULONG);
 static NTSTATUS (WINAPI *pNtMapViewOfSectionEx)(HANDLE, HANDLE, PVOID *, const LARGE_INTEGER *, SIZE_T *,
         ULONG, ULONG, MEM_EXTENDED_PARAMETER *, ULONG);
+static NTSTATUS (WINAPI *pNtSetInformationVirtualMemory)(HANDLE, VIRTUAL_MEMORY_INFORMATION_CLASS,
+                                                         ULONG_PTR, PMEMORY_RANGE_ENTRY,
+                                                         PVOID, ULONG);
 
 static const BOOL is_win64 = sizeof(void*) != sizeof(int);
 static BOOL is_wow64;
@@ -1279,6 +1282,165 @@ static void test_syscalls(void)
     UnmapViewOfFile( ptr );
 }
 
+static void test_prefetch(void)
+{
+    NTSTATUS status;
+    MEMORY_RANGE_ENTRY entries[2] = {{ 0 }};
+    ULONG reservedarg = 0;
+    char stackmem[] = "Test stack mem";
+    static char testmem[] = "Test memory range data";
+
+    if (!pNtSetInformationVirtualMemory)
+    {
+        skip("no NtSetInformationVirtualMemory in ntdll\n");
+        return;
+    }
+
+    status = pNtSetInformationVirtualMemory( NtCurrentProcess(), -1UL, 1, entries, NULL, 32);
+    ok( status == STATUS_INVALID_PARAMETER_2,
+        "NtSetInformationVirtualMemory unexpected status on invalid info class (1): %08x\n", status);
+
+    status = pNtSetInformationVirtualMemory( NtCurrentProcess(), -1UL, 0, NULL, NULL, 0);
+    ok( status == STATUS_INVALID_PARAMETER_2 || (is_wow64 && status == STATUS_INVALID_PARAMETER_3),
+        "NtSetInformationVirtualMemory unexpected status on invalid info class (2): %08x\n", status);
+    if (is_wow64)
+    {
+        todo_wine
+        ok( status == STATUS_INVALID_PARAMETER_3,
+            "wow64 NtSetInformationVirtualMemory unexpected status on invalid info class (2): %08x\n", status);
+    }
+
+    status = pNtSetInformationVirtualMemory( NtCurrentProcess(), -1UL, 1, NULL, NULL, 32);
+    ok( status == STATUS_INVALID_PARAMETER_2 || (is_wow64 && status == STATUS_ACCESS_VIOLATION),
+        "NtSetInformationVirtualMemory unexpected status on invalid info class (3): %08x\n", status);
+    if (is_wow64)
+    {
+        todo_wine
+        ok( status == STATUS_ACCESS_VIOLATION,
+            "wow64 NtSetInformationVirtualMemory unexpected status on invalid info class (3): %08x\n", status);
+    }
+
+    status = pNtSetInformationVirtualMemory( NtCurrentProcess(), VmPrefetchInformation,
+                                             1, entries, NULL, 0 );
+    ok( status == STATUS_INVALID_PARAMETER_5 ||
+        broken( is_wow64 && status == STATUS_INVALID_PARAMETER_6 ) /* win10 1507 */,
+        "NtSetInformationVirtualMemory unexpected status on NULL info data (1): %08x\n", status);
+
+    status = pNtSetInformationVirtualMemory( NtCurrentProcess(), VmPrefetchInformation,
+                                             1, NULL, NULL, 0 );
+    ok( status == STATUS_INVALID_PARAMETER_5 || (is_wow64 && status == STATUS_ACCESS_VIOLATION),
+        "NtSetInformationVirtualMemory unexpected status on NULL info data (2): %08x\n", status);
+    if (is_wow64)
+    {
+        todo_wine
+        ok( status == STATUS_ACCESS_VIOLATION,
+            "wow64 NtSetInformationVirtualMemory unexpected status on NULL info data (2): %08x\n", status);
+    }
+
+    status = pNtSetInformationVirtualMemory( NtCurrentProcess(), VmPrefetchInformation,
+                                             0, NULL, NULL, 0 );
+    ok( status == STATUS_INVALID_PARAMETER_5 || (is_wow64 && status == STATUS_INVALID_PARAMETER_3),
+        "NtSetInformationVirtualMemory unexpected status on NULL info data (3): %08x\n", status);
+    if (is_wow64)
+    {
+        todo_wine
+        ok( status == STATUS_INVALID_PARAMETER_3,
+            "wow64 NtSetInformationVirtualMemory unexpected status on NULL info data (3): %08x\n", status);
+    }
+
+    status = pNtSetInformationVirtualMemory( NtCurrentProcess(), VmPrefetchInformation,
+                                             1, entries, &reservedarg, sizeof(reservedarg) * 2 );
+    ok( status == STATUS_INVALID_PARAMETER_6,
+        "NtSetInformationVirtualMemory unexpected status on extended info data (1): %08x\n", status);
+
+    status = pNtSetInformationVirtualMemory( NtCurrentProcess(), VmPrefetchInformation,
+                                             0, NULL, &reservedarg, sizeof(reservedarg) * 2 );
+    ok( status == STATUS_INVALID_PARAMETER_6 || (is_wow64 && status == STATUS_INVALID_PARAMETER_3),
+        "NtSetInformationVirtualMemory unexpected status on extended info data (2): %08x\n", status);
+    if (is_wow64)
+    {
+        todo_wine
+        ok( status == STATUS_INVALID_PARAMETER_3,
+            "wow64 NtSetInformationVirtualMemory unexpected status on extended info data (2): %08x\n", status);
+    }
+
+    status = pNtSetInformationVirtualMemory( NtCurrentProcess(), VmPrefetchInformation,
+                                             1, entries, &reservedarg, sizeof(reservedarg) / 2 );
+    ok( status == STATUS_INVALID_PARAMETER_6,
+        "NtSetInformationVirtualMemory unexpected status on shrunk info data (1): %08x\n", status);
+
+    status = pNtSetInformationVirtualMemory( NtCurrentProcess(), VmPrefetchInformation,
+                                             0, NULL, &reservedarg, sizeof(reservedarg) / 2 );
+    ok( status == STATUS_INVALID_PARAMETER_6 || (is_wow64 && status == STATUS_INVALID_PARAMETER_3),
+        "NtSetInformationVirtualMemory unexpected status on shrunk info data (2): %08x\n", status);
+    if (is_wow64)
+    {
+        todo_wine
+        ok( status == STATUS_INVALID_PARAMETER_3,
+            "wow64 NtSetInformationVirtualMemory unexpected status on shrunk info data (2): %08x\n", status);
+    }
+
+    status = pNtSetInformationVirtualMemory( NtCurrentProcess(), VmPrefetchInformation,
+                                             0, NULL, &reservedarg, sizeof(reservedarg) );
+    ok( status == STATUS_INVALID_PARAMETER_3,
+        "NtSetInformationVirtualMemory unexpected status on 0 entries: %08x\n", status);
+
+    status = pNtSetInformationVirtualMemory( NtCurrentProcess(), VmPrefetchInformation,
+                                             1, NULL, &reservedarg, sizeof(reservedarg) );
+    ok( status == STATUS_ACCESS_VIOLATION,
+        "NtSetInformationVirtualMemory unexpected status on NULL entries: %08x\n", status);
+
+    entries[0].VirtualAddress = NULL;
+    entries[0].NumberOfBytes = 0;
+    status = pNtSetInformationVirtualMemory( NtCurrentProcess(), VmPrefetchInformation,
+                                             1, entries, &reservedarg, sizeof(reservedarg) );
+    ok( status == STATUS_INVALID_PARAMETER_4 ||
+        broken( is_wow64 && status == STATUS_INVALID_PARAMETER_6 ) /* win10 1507 */,
+        "NtSetInformationVirtualMemory unexpected status on 1 empty entry: %08x\n", status);
+
+    entries[0].VirtualAddress = NULL;
+    entries[0].NumberOfBytes = page_size;
+    status = pNtSetInformationVirtualMemory( NtCurrentProcess(), VmPrefetchInformation,
+                                             1, entries, &reservedarg, sizeof(reservedarg) );
+    ok( status == STATUS_SUCCESS ||
+        broken( is_wow64 && status == STATUS_INVALID_PARAMETER_6 ) /* win10 1507 */,
+        "NtSetInformationVirtualMemory unexpected status on 1 NULL address entry: %08x\n", status);
+
+    entries[0].VirtualAddress = ULongToPtr(PtrToUlong(testmem) & -(ULONG_PTR)page_size);
+    entries[0].NumberOfBytes = page_size;
+    status = pNtSetInformationVirtualMemory( NtCurrentProcess(), VmPrefetchInformation,
+                                             1, entries, &reservedarg, sizeof(reservedarg) );
+    ok( status == STATUS_SUCCESS ||
+        broken( is_wow64 && status == STATUS_INVALID_PARAMETER_6 ) /* win10 1507 */,
+        "NtSetInformationVirtualMemory unexpected status on 1 page-aligned entry: %08x\n", status);
+
+    entries[0].VirtualAddress = testmem;
+    entries[0].NumberOfBytes = sizeof(testmem);
+    status = pNtSetInformationVirtualMemory( NtCurrentProcess(), VmPrefetchInformation,
+                                             1, entries, &reservedarg, sizeof(reservedarg) );
+    ok( status == STATUS_SUCCESS ||
+        broken( is_wow64 && status == STATUS_INVALID_PARAMETER_6 ) /* win10 1507 */,
+        "NtSetInformationVirtualMemory unexpected status on 1 entry: %08x\n", status);
+
+    entries[0].VirtualAddress = NULL;
+    entries[0].NumberOfBytes = page_size;
+    status = pNtSetInformationVirtualMemory( NtCurrentProcess(), VmPrefetchInformation,
+                                             1, entries, &reservedarg, sizeof(reservedarg) );
+    ok( status == STATUS_SUCCESS ||
+        broken( is_wow64 && status == STATUS_INVALID_PARAMETER_6 ) /* win10 1507 */,
+        "NtSetInformationVirtualMemory unexpected status on 1 unmapped entry: %08x\n", status);
+
+    entries[0].VirtualAddress = ULongToPtr(PtrToUlong(testmem) & -(ULONG_PTR)page_size);
+    entries[0].NumberOfBytes = page_size;
+    entries[1].VirtualAddress = ULongToPtr(PtrToUlong(stackmem) & -(ULONG_PTR)page_size);
+    entries[1].NumberOfBytes = page_size;
+    status = pNtSetInformationVirtualMemory( NtCurrentProcess(), VmPrefetchInformation,
+                                             2, entries, &reservedarg, sizeof(reservedarg) );
+    ok( status == STATUS_SUCCESS ||
+        broken( is_wow64 && status == STATUS_INVALID_PARAMETER_6 ) /* win10 1507 */,
+        "NtSetInformationVirtualMemory unexpected status on 2 page-aligned entries: %08x\n", status);
+}
+
 START_TEST(virtual)
 {
     HMODULE mod;
@@ -1308,6 +1470,7 @@ START_TEST(virtual)
     pRtlGetEnabledExtendedFeatures = (void *)GetProcAddress(mod, "RtlGetEnabledExtendedFeatures");
     pNtAllocateVirtualMemoryEx = (void *)GetProcAddress(mod, "NtAllocateVirtualMemoryEx");
     pNtMapViewOfSectionEx = (void *)GetProcAddress(mod, "NtMapViewOfSectionEx");
+    pNtSetInformationVirtualMemory = (void *)GetProcAddress(mod, "NtSetInformationVirtualMemory");
 
     NtQuerySystemInformation(SystemBasicInformation, &sbi, sizeof(sbi), NULL);
     trace("system page size %#lx\n", sbi.PageSize);
@@ -1320,4 +1483,5 @@ START_TEST(virtual)
     test_NtMapViewOfSectionEx();
     test_user_shared_data();
     test_syscalls();
+    test_prefetch();
 }

-- 
2.36.0

